src/main/java/se/liu/ida/hefquin/engine/queryplan/executable/impl/ops/ExecOpRequestTPFWithTranslation.java

```diff
@@ -0,0 +1,202 @@
+			resList.addAll(handleBGP((BGP) reqTranslation, fedAccessMgr));
+		}
+		
+		else {
+			throw new FederationAccessException(reqTranslation.toString(), req, fm);
+		}
+		
+		return new TPFResponseImpl(resList, resList, null, fm, req, startTime);
+	}
+	
+	protected Set<Triple> handleTriplePattern(final TriplePattern tp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		final TPFRequest newReq = new TPFRequestImpl(tp);
+		TPFResponse res;
+		if (fm instanceof BRTPFServer) {
+			res = FederationAccessUtils.performRequest(fedAccessMgr, newReq, (BRTPFServer) fm);
+		} else if (fm instanceof TPFServer) {
+			res = FederationAccessUtils.performRequest(fedAccessMgr, newReq, (TPFServer) fm);
+		} else {
+			throw new FederationAccessException(fm.toString(), req, fm);
+		}
+		final Set<Triple> tripleTranslation = new HashSet<>();
+		for(final Triple i : res.getPayload()) {
+			tripleTranslation.addAll(translateResultTriple(i, newReq.getQueryPattern()));
+		}
+		return tripleTranslation;
+	}
+	
+	protected Set<Triple> handleUnionPattern(final SPARQLUnionPattern up, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		final Set<Triple> unionTranslation = new HashSet<>();
+		for(final SPARQLGraphPattern i : up.getSubPatterns()) {
+			if (i instanceof TriplePattern) {
+				unionTranslation.addAll(handleTriplePattern(((TriplePattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLUnionPattern) {
+				unionTranslation.addAll(handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLGroupPattern) {
+				unionTranslation.addAll(handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr));
+			} else if (i instanceof BGP) {
+				unionTranslation.addAll(handleBGP(((BGP) i), fedAccessMgr));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+		}
+		return unionTranslation;
+	}
+	
+	protected Set<Triple> handleGroupPattern(final SPARQLGroupPattern gp, final FederationAccessManager fedAccessMgr) throws FederationAccessException{
+		Set<Triple> groupTranslation = null;
+		for(final SPARQLGraphPattern i : gp.getSubPatterns()) {
+			final Set<Triple> partialTranslation = new HashSet<>();
+			if (i instanceof TriplePattern) {
+				partialTranslation.addAll(handleTriplePattern((TriplePattern) i, fedAccessMgr));
+			} else if (i instanceof SPARQLUnionPattern) {
+				partialTranslation.addAll(handleUnionPattern(((SPARQLUnionPattern) i), fedAccessMgr));
+			} else if (i instanceof SPARQLGroupPattern) {
+				partialTranslation.addAll(handleGroupPattern(((SPARQLGroupPattern) i), fedAccessMgr));
+			} else if (i instanceof BGP) {
+				partialTranslation.addAll(handleBGP(((BGP) i), fedAccessMgr));
+			} else {
+				throw new FederationAccessException(i.toString(), req, fm);
+			}
+			
+			if (groupTranslation == null) {
+				groupTranslation = new HashSet<>();
+				groupTranslation.addAll(partialTranslation);
```

Comment:
Why not initialize the HashSet outside the for loop (this way, groupTranslation can be made final), and here check if isEmpty()

Reply:
I think I did it this way, in case the retain all at some point leads to the results being empty. Is my thought process correct or should I change it?